src/Voron/Impl/Journal/JournalReader.cs

```diff
@@ -237,11 +243,76 @@ public bool ReadOneTransactionToDataFile(ref Pager.State dataPagerState, ref Pag
                 NativeMemory.Free(currentBuffer);
             }
 
+            if ((current->Flags & TransactionPersistenceModeFlags.HasFreePages) == TransactionPersistenceModeFlags.HasFreePages)
+            {
+                var freePages = ReadEncodedFreePage(outputPage + totalRead, _allocator);
+
+                foreach (var freedPage in freePages)
+                {
+                    _modifiedPages.Remove(freedPage);
+                }
+            }
+            
             LastTransactionHeader = current;
 
             return true;
         }
 
+        internal static List<long> ReadEncodedFreePage(byte* src, ByteStringContext allocator)
+        {
+            var readPtr = src;
+
+            var headerPtr = readPtr;
+
+            var header = *(FreePagesHeader*)headerPtr;
+            var sectionHeaders = (EncodedFreePagesSection*)(headerPtr + sizeof(FreePagesHeader) + header.EncodedSectionsSize);
+
+            readPtr += sizeof(FreePagesHeader);
+            
+            var result = new List<long>(header.NumberOfPages);
+            
+            var readBatchSize = 256;
+            
+            using var _ = allocator.Allocate(readBatchSize * sizeof(long), out var buffer);
+                
+            long* freePages = (long*)buffer.Ptr;
+
+            var totalCount = 0;
+            
+            for (int i = 0; i < header.EncodedSectionsCount; i++)
+            {
+                var sectionSize = sectionHeaders[i].Size;
+                
+                using var reader = new FastPForBufferedReader(allocator);
```

Comment:
Does this free all the memory?
Do we need to *reset* the allocator between journals?